### ISSUE TITLE
Future-Proofing Images with Subs

### ISSUE THREAD

**Issue by Alexhuszagh** [cross-rs/cross#1020]
Rust triples are allowed to contain `.` characters, although I believe this is limited only to the architecture portion of the triple (and not the OS, vendor, or environment). For example, `thumbv8m.main-none-eabihf` is a valid LLVM triple. This means that splitting on a `.` character isn't necessarily going to give us our desired sub, and might produce erroneous results in the future.

We can get a bit more clarity looking at the comments in the [llvm/ADT/Triple.h](https://llvm.org/doxygen/Triple_8h_source.html) header. First, we want too look into the [parser](https://llvm.org/doxygen/Triple_8cpp_source.html#l00941), in `std::string Triple::normalize(StringRef Str)`. This isn't too bad, since it splits off into `parseArch`, `parseSubArch`, `parseVendor`, `parseOS`, and `parseEnvironment`.

We can see `parseArch` has some known values, and special cases for ARM and BPF-family architectures. The vendors are hard-coded and only accepted ASCII lowercase letters. The OSes and environments only expect strings **starting** with ASCII numbers and letters,  but are flexible for any possible characters (other than `-`) at the end of each identifier. The architecture field, however, will accept `.` characters for a few targets.

In short, the following code isn't generalizable enough: this will cause an invalid target/sub to be parsed from some valid Rust triples:

https://github.com/cross-rs/cross/blob/7c06ad9f8b195d2a1867b176bf9335721f3186cd/xtask/src/util.rs#L164-L180
